fdbserver/DataDistribution.actor.cpp

```diff
@@ -4860,7 +5199,20 @@ ACTOR Future<Void> storageRecruiter(DDTeamCollection* self,
 						    .detail("NumExistingSS", numExistingSS);
 					}
 
-					if (hasHealthyTeam && !tssState->active && tssToRecruit > 0) {
+					// check whether this candidate is a process under perpetual wiggling, but be paused because of
```

Comment:
Is this comment + logic for the case where the cluster is unhealthy, we don't want to replace a storage server that was wiggled out with a TSS because we need more storage servers in the cluster to return to healthy status?
If that can cause problems, we may also need to take this into account in the later phases of TSS recruitment.
For example, if the TSS recruitment cycle is currently holding a TSS process and waiting for an SS to pair it with, but the wiggle gets paused due to an unhealthy cluster, we may want to cancel the hold on that TSS process and recruit it as a storage server. This would be similar to the logic where we already cancel that if the desired tss count is negative or there are zero healthy teams in the cluster.

Reply:
Also, my TSS mapping change is also going to have conflicts with this :(
But I think we should get this merged first and I can rebase my changes after.

Reply:
My proposed scenario is like the storage server candidate is the server under wiggling. Suppose when half of the data has been moved to another team, the wiggler find the cluster was not healthy and paused the wiggling process. Now the candidate has half of the data remaining. If it becomes a TSS or TSS pair, I think the remained data cannot reach the required replica factor. (And we also want a TSS is clean)

Reply:
I think the distinction here is that the cluster controller shouldn't return a storage process as a candidate, unless it doesn't currently have any storage processes on it at all. So if half of the data was moved away, there would still be a storage process on that worker.
But if for some reason a worker with an existing storage process was returned as a candidate, the initialize storage request (TSS or normal) would create a new, empty, storage server/TSS on that worker, it wouldn't replace the old one or convert it to/from a TSS.
So I don't think it is possible for this to happen.

Reply:
Yes, it makes sense. It will let the code more simple, thank you for pointing this out.